### PR TITLE
Ecal DPG - Backporting new software[PR 27175] to recover isolated new channels

### PR DESCRIPTION
**Pull request by rchatter**
Backporting PR#27175 [@nancymarinelli @argiro ]

In order to validate and backport PR28040 we need to backport PR#27175 and PR#28514 to make a validation for which PPD will have to advice us on the way how to do it.

This is the first step where we are backporting PR#27175 [ https://github.com/cms-sw/cmssw/pull/27175], which has already been merged.

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by rchatter** [comment]
Sorry for this. Will investigate and re-commit.

**Comment by slava77** [comment]
@cmsbuild please test with cms-sw/cmsdist#5516

File: RecoLocalCalo/EcalRecProducers/python/ecalRecHit_cfi.py [hunk 1]
```diff
@@ -49,7 +49,7 @@
                             
     # for channel recovery
     algoRecover = cms.string("EcalRecHitWorkerRecover"),
-    recoverEBIsolatedChannels = cms.bool(False),
+                            recoverEBIsolatedChannels = cms.bool(True),##default is false
```
Comment (perrotta) [inline RecoLocalCalo/EcalRecProducers/python/ecalRecHit_cfi.py]: Please keep the defaut "false" for 10_6 (and indent correctly)

Comment (perrotta) [reply RecoLocalCalo/EcalRecProducers/python/ecalRecHit_cfi.py]: This is quite likely the reason of the differences observed in the test outpus

**Review by perrotta — COMMENTED** [review]
Several files have been directly copied from the 11_X version: this brings together all clang formatting changes that were applied in the meanwhile, unrelated to this PR but not harming for the backport (as I checked Several files have been directly copied from the 11_X version: this brings together all clang formatting changes that were applied in the meanwhile, unrelated to this PR but not harming (as I checked Several files have been directly copied from the 11_X version: this brings together all clang formatting changes that were applied in the meanwhile, unrelated to this PR but not harming (as I checked thoroughly)

File: RecoLocalCalo/EcalRecProducers/python/ecalRecHit_cfi.py (hunk 1 above)
Comment (rchatter) [reply RecoLocalCalo/EcalRecProducers/python/ecalRecHit_cfi.py]: I see. Done.

File: RecoLocalCalo/EcalDeadChannelRecoveryAlgos/interface/EcalDeadChannelRecoveryBDTG.h [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+private:
+  const CaloTopology *topology_;
+  struct XtalMatrix {
+    std::array<float, 9> rEn, ieta, iphi;
+    float sumE8;
+  };
+
+  XtalMatrix mx_;
+
+  edm::FileInPath bdtWeightFileNoCracks_;
```
[the remark this answers is not in the document]
Comment (rchatter) [reply RecoLocalCalo/EcalDeadChannelRecoveryAlgos/interface/EcalDeadChannelRecoveryBDTG.h]: Done.

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
@cmsbuild please test with cms-sw/cmsdist#5516

File: RecoLocalCalo/EcalDeadChannelRecoveryAlgos/interface/EcalDeadChannelRecoveryAlgos.h [hunk 3]
```diff
@@ -7,16 +7,25 @@
...
-                     std::string algo, double Sum8Cut, bool *AccFlag);
+  float correct(const DetIdT id,
+                const EcalRecHitCollection &hit_collection,
+                std::string algo,
+                double single8Cut,
+                double sum8Cut,
+                bool *accFlag);
 
- private:
-  EcalDeadChannelRecoveryNN<DetIdT> nn;
```
Comment (perrotta) [inline RecoLocalCalo/EcalDeadChannelRecoveryAlgos/interface/EcalDeadChannelRecoveryAlgos.h]: This is just to be sure: backports into a production release must satisfy the "no change" rule. Here you remove the NN method for the deadchannel recovery, and substitute it with the BDTG one.
The default setting for ```recoverEBIsolatedChannels``` is ```false```, and I don't see in CMSSW any place or official workflow in which it is turned into ```true```. Therefore, I assume that the backport cannot modify the result in any of those official workflows.
Can you confirm that there are no use cases in 10_6_X that require such NN method for the ecal deadchannel recovery, and it id therefore safe to fully replace it with the BDTG one?

**Comment by nancymarinelli** [comment]
On 05/02/20 09:57, perrotta wrote:
> perrotta commented on this pull request.
>
>
>
>>   
> - private:
> -  EcalDeadChannelRecoveryNN<DetIdT> nn;
>
> This is just to be sure: backports into a production release must satisfy the "no change" rule. Here you remove the NN method for the deadchannel recovery, and substitute it with the BDTG one.
> The default setting for ```recoverEBIsolatedChannels``` is ```false```, and I don't see in CMSSW any place or official workflow in which it is turned into ```true```. Therefore, I assume that the backport cannot modify the result in any of those official workflows.
No modification of the official workflows. A special configuration will 
be made in due time, according to plans with PPD for the large 
validation scale
> Can you confirm that there are no use cases in 10_6_X that require such NN method for the ecal deadchannel recovery, and it id therefore safe to fully replace it with the BDTG one?
>
Hi, no use cases for NN.


-- 
________________________________________
Nancy Marinelli
Research Associate Professor
University of Notre Dame, IN, US
CERN, Bdg 40/3-A01, 1211 Geneva
SWITZERLAND
Phone +41-22-76-70809
fax +41-22-76-78940

**Comment by perrotta** [comment]
+1
- Backport of #27175, and the subsequent code optimizations implemented in master for the same
- It replaces the previous Ecal dead channel recovery method based on NN with a new one based on BDT 
  - Ecal dead channel recovery is not enabled in any standard or used workflow, therefore removing the NN based recovery method does not affect the output of any such workflow in 10_6
- Several updated files have been directly copied from the 11_X version: this brings together all clang formatting changes that were applied in the meanwhile, unrelated to this PR. This is not harming and it does not modify the flow of the code, as I checked explicitely
- Jenkins tests pass and show no differences ,as expected

**Comment by silviodonato** [comment]
code-checks

File: RecoLocalCalo/EcalDeadChannelRecoveryAlgos/interface/EcalDeadChannelRecoveryBDTG.h [hunk 4]
```diff
@@ -0,0 +1,51 @@
+#ifndef RecoLocalCalo_EcalDeadChannelRecoveryAlgos_EcalDeadChannelRecoveryBDTG_H
+#define RecoLocalCalo_EcalDeadChannelRecoveryAlgos_EcalDeadChannelRecoveryBDTG_H
+
```
Comment (silviodonato) [inline RecoLocalCalo/EcalDeadChannelRecoveryAlgos/interface/EcalDeadChannelRecoveryBDTG.h]: See https://github.com/cms-sw/cmssw/commit/ffabcf01891076f6687bad559f7017e1079f9339#diff-4524ebcfc3b2ca5e718d8a00ddbe71a9R4

File: RecoLocalCalo/EcalRecProducers/plugins/EcalRecHitProducer.cc [hunk 5]
```diff
@@ -27,298 +27,294 @@
...
+
+  eeDetIdToBeRecoveredToken_ = consumes<std::set<EEDetId>>(ps.getParameter<edm::InputTag>("eeDetIdToBeRecovered"));
+
+  ebFEToBeRecoveredToken_ = consumes<std::set<EcalTrigTowerDetId>>(ps.getParameter<edm::InputTag>("ebFEToBeRecovered"));
+
+  eeFEToBeRecoveredToken_ = consumes<std::set<EcalScDetId>>(ps.getParameter<edm::InputTag>("eeFEToBeRecovered"));
+
+  std::string componentType = ps.getParameter<std::string>("algo");
+  edm::ConsumesCollector c{consumesCollector()};
+  worker_ = std::unique_ptr<EcalRecHitWorkerBaseClass>{EcalRecHitWorkerFactory::get()->create(componentType, ps, c)};
```
Comment (silviodonato) [inline RecoLocalCalo/EcalRecProducers/plugins/EcalRecHitProducer.cc]: https://github.com/cms-sw/cmssw/commit/ffabcf01891076f6687bad559f7017e1079f9339#diff-d95390db2a2c8030770f2749c29d2f37R61

File: RecoLocalCalo/EcalRecProducers/plugins/EcalRecHitProducer.cc [hunk 6]
```diff
@@ -27,298 +27,294 @@
...
+
+  // to recover problematic channels
+  componentType = ps.getParameter<std::string>("algoRecover");
+  workerRecover_ = std::unique_ptr<EcalRecHitWorkerBaseClass>{EcalRecHitWorkerFactory::get()->create(componentType, ps, c)};
```
Comment (silviodonato) [inline RecoLocalCalo/EcalRecProducers/plugins/EcalRecHitProducer.cc]: See https://github.com/cms-sw/cmssw/commit/ffabcf01891076f6687bad559f7017e1079f9339#diff-d95390db2a2c8030770f2749c29d2f37R65

**Review by silviodonato — COMMENTED** [review]
I found three small differences between #28831 and #27175.
Could you confirm they are harmless?

**Comment by silviodonato** [comment]
+1
all differences were simply cleaning and code-checks changes.

**Comment by Martin-Grunewald** [comment]
Seems this PR messes up many 10_6 IB relvals.

**Comment by Martin-Grunewald** [comment]
```
----- Begin Fatal Exception 10-Feb-2020 10:35:04 CET-----------------------
An exception of category 'FileInPathError' occurred while
   [0] Constructing the EventProcessor
Exception Message:
edm::FileInPath unable to find file RecoLocalCalo/EcalDeadChannelRecoveryAlgos/\
data/BDTWeights/bdtgAllRH_8GT700MeV_noCracks_ZskimData2017_v1.xml anywhere in t\
he search path.
The search path is defined by: CMSSW_SEARCH_PATH
${CMSSW_SEARCH_PATH} is: /data/CMS6/CMSSW_10_6_X_2020-02-09-0000/poison:/data/C\
MS6/CMSSW_10_6_X_2020-02-09-0000/src:/data/CMS6/CMSSW_10_6_X_2020-02-09-0000/ex\
ternal/slc7_amd64_gcc700/data:/cvmfs/cms-ib.cern.ch/week1/slc7_amd64_gcc700/cms\
/cmssw/CMSSW_10_6_X_2020-02-09-0000/src:/cvmfs/cms-ib.cern.ch/week1/slc7_amd64_\
gcc700/cms/cmssw/CMSSW_10_6_X_2020-02-09-0000/external/slc7_amd64_gcc700/data
Current directory is: /data/CMS6/CMSSW_10_6_X_2020-02-09-0000/src/HLTrigger/Con\
figuration/test
----- End Fatal Exception -------------------------------------------------
```

**Comment by perrotta** [comment]
As I wrote in the [approval notes](https://www.reggionline.com/albinea-svaligiato-nella-notte-negozio-biciclette-maxent/), this PR requires an external: cms-sw/cmsdist#5516
Please @silviodonato also merge that one

**Comment by silviodonato** [comment]
Right, I have merged them right now. Thanks @Martin-Grunewald @perrotta

**Pull request merged**